core/tag.c

```diff
@@ -39,20 +39,43 @@ tag_t lf_tag(void* env) {
   return ((environment_t*)env)->current_tag;
 }
 
+instant_t lf_time_add(instant_t a, interval_t b) {
+  if (a == NEVER || b == NEVER) {
+    return NEVER;
+  }
+  if (a == FOREVER || b == FOREVER) {
+    return FOREVER;
+  }
+  instant_t res = a + b;
+  // Check for overflow
+  if (res < a && b > 0) {
+    return FOREVER;
+  }
+  // Check for underflow
+  if (res > a && b < 0) {
+    return NEVER;
+  }
+  return res;
+}
+
 tag_t lf_tag_add(tag_t a, tag_t b) {
-  if (a.time == NEVER || b.time == NEVER)
-    return NEVER_TAG;
-  if (a.time == FOREVER || b.time == FOREVER)
+  instant_t res = lf_time_add(a.time, b.time);
+  if (res == FOREVER) {
     return FOREVER_TAG;
-  if (b.time > 0)
+  }
+  if (res == NEVER) {
+    return NEVER_TAG;
+  }
+
+  if (b.time > 0) {
     a.microstep = 0; // Ignore microstep of first arg if time of second is > 0.
-  tag_t result = {.time = a.time + b.time, .microstep = a.microstep + b.microstep};
-  if (result.microstep < a.microstep)
-    return FOREVER_TAG;
-  if (result.time < a.time && b.time > 0)
+  }
+  tag_t result = {.time = res, .microstep = a.microstep + b.microstep};
+
+  // If microsteps overflows
+  if (result.microstep < a.microstep) {
```

Comment:
Why are we returning `FOREVER_TAG` if the microstep overflows? Is it even a problem for the microstep to overflow? If it is a problem, that must mean there is an event on the queue that has a greater tag while it is supposed to denote an earlier event. But before this happens, wouldn't we likely have run out of memory already? I'm not quite sure what the right answer is, but the current choice strikes me as odd...

Reply:
It does seem like something that should trigger a runtime error instead. Not sure. Could this happen any other way than a action scheduled with 0 delay 2^32 times?

Reply:
I don't think it can be achieved in any other way. But if that's what happens, then wrapping around is a fine solution imo, and not an error. Moving time to infinity in response seems like a very weird thing to do.

Reply:
If it wraps around it would be an error situation. If you just wrap around then the new tag is less than the previous. This could result in scheduling a logical action into the past, and the event will just be discarded. The other, more reasonable alternative in my opinion is to increment the timestamp. This way, at least the event is not dropped. However semantically it would be incorrect. I think in both cases an error must be printed

Reply:
I will create an issue for this and reference it here in the code